### ISSUE TITLE
Build failure with Intel & GCC compiler

### ISSUE THREAD

**Issue by pramodk** [arbor-sim/arbor#719]
I may not have read documentation thoroughly and building arbor first time.  I am using the scripts from [nsuite](https://github.com/arbor-sim/nsuite) which worked fine on Daint. But when I build on another cluster I got following errors : 

* Using Intel 18 : 

```
HEAD is now at 6f70bfe... Bump version number to v0.2
-- The C compiler identification is Intel 18.0.1.20171018
-- The CXX compiler identification is Intel 18.0.1.20171018
-- Check for working C compiler: /opt/hpc/mpt/mpt-2.16/bin/mpicc
-- Check for working C compiler: /opt/hpc/mpt/mpt-2.16/bin/mpicc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /opt/hpc/mpt/mpt-2.16/bin/mpicxx
-- Check for working CXX compiler: /opt/hpc/mpt/mpt-2.16/bin/mpicxx -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
....
[ 50%] Building CXX object arbor/CMakeFiles/arbor.dir/memory/util.cpp.o
In file included from /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/util/uninitialized.hpp(15),
                 from /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/util/optional.hpp(43),
                 from /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/util/maputil.hpp(9),
                 from /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/mechcat.cpp(9):
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/util/compat.hpp(50): warning #3175: unrecognized gcc optimization level
  __attribute((optimize("no-tree-vectorize")))
                        ^

[ 50%] Building CXX object arbor/CMakeFiles/arbor.dir/merge_events.cpp.o
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/mechcat.cpp(138): error: call of an object of a class type without appropriate operator() or conversion functions to pointer-to-function type
              self(self, p->parent, mptr);
              ^
          detected during instantiation of function "lambda [](auto &, const std::__cxx11::string &, arb::mechanism *)->void [with <auto-1>=lambda [](auto &, const std::__cxx11::string &, arb::mechanism *)->void]" at line 145

compilation aborted for /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/mechcat.cpp (code 2)
make[3]: *** [arbor/CMakeFiles/arbor.dir/mechcat.cpp.o] Error 2
make[3]: *** Waiting for unfinished jobs....
In file included from /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/math.hpp(8),
                 from /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/merge_events.cpp(7):
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/util/compat.hpp(50): warning #3175: unrecognized gcc optimization level
  __attribute((optimize("no-tree-vectorize")))
```

* Using gcc 7.3 : 

```
-- The C compiler identification is GNU 7.3.0
-- The CXX compiler identification is GNU 7.3.0
-- Check for working C compiler: /opt/hpc/mpt/mpt-2.16/bin/mpicc
-- Check for working C compiler: /opt/hpc/mpt/mpt-2.16/bin/mpicc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /opt/hpc/mpt/mpt-2.16/bin/mpicxx
-- Check for working CXX compiler: /opt/hpc/mpt/mpt-2.16/bin/mpicxx -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE
-- Found MPI_CXX: /opt/hpc/mpt/mpt-2.16/bin/mpicxx (found version "3.1")
-- Found MPI: TRUE (found version "3.1") found components:  CXX
-- Could NOT find Unwind (missing: Unwind_LIBRARIES)
-- Found Sphinx: /usr/bin/sphinx-build

[ 90%] Building CXX object example/brunel/CMakeFiles/brunel-miniapp.dir/brunel_miniapp.cpp.o
[ 92%] Building CXX object example/brunel/CMakeFiles/brunel-miniapp.dir/io.cpp.o
In file included from /local/scratch/tmp/ARBOR/nsuite/build/arbor/example/bench/bench.cpp:22:0:
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arborenv/include/arborenv/gpu_env.hpp:8:5: error: ‘int arbenv::find_private_gpu(Comm) [with Comm = <unnamed enum>]’, declared using unnamed type, is used but never defined [-fpermissive]
 int find_private_gpu(Comm comm);
     ^~~~~~~~~~~~~~~~
In file included from /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/profile/meter_manager.hpp:7:0,
                 from /local/scratch/tmp/ARBOR/nsuite/build/arbor/example/bench/bench.cpp:11:
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/context.hpp:62:9: error: ‘arb::context arb::make_context(const arb::proc_allocation&, Comm) [with Comm = <unnamed enum>; arb::context = std::unique_ptr<arb::execution_context, void (*)(arb::execution_context*)>]’, declared using unnamed type, is used but never defined [-fpermissive]
 context make_context(const proc_allocation& resources, Comm comm);
         ^~~~~~~~~~~~
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/context.hpp:62:9: warning: ‘arb::context arb::make_context(const arb::proc_allocation&, Comm) [with Comm = <unnamed enum>]’ used but never defined
In file included from /local/scratch/tmp/ARBOR/nsuite/build/arbor/example/bench/bench.cpp:22:0:
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arborenv/include/arborenv/gpu_env.hpp:8:5: warning: ‘int arbenv::find_private_gpu(Comm) [with Comm = <unnamed enum>]’ used but never defined
 int find_private_gpu(Comm comm);
     ^~~~~~~~~~~~~~~~
make[3]: *** [example/bench/CMakeFiles/bench.dir/bench.cpp.o] Error 1
make[2]: *** [example/bench/CMakeFiles/bench.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
In file included from /local/scratch/tmp/ARBOR/nsuite/build/arbor/example/brunel/brunel_miniapp.cpp:23:0:
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arborenv/include/arborenv/gpu_env.hpp:8:5: error: ‘int arbenv::find_private_gpu(Comm) [with Comm = <unnamed enum>]’, declared using unnamed type, is used but never defined [-fpermissive]
 int find_private_gpu(Comm comm);
     ^~~~~~~~~~~~~~~~
```

* Using GCC 6.3 

```
-- The C compiler identification is GNU 6.4.0
-- The CXX compiler identification is GNU 6.4.0
-- Check for working C compiler: /opt/hpc/mpt/mpt-2.16/bin/mpicc
-- Check for working C compiler: /opt/hpc/mpt/mpt-2.16/bin/mpicc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /opt/hpc/mpt/mpt-2.16/bin/mpicxx
-- Check for working CXX compiler: /opt/hpc/mpt/mpt-2.16/bin/mpicxx -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE
-- Found MPI_CXX: /opt/hpc/mpt/mpt-2.16/bin/mpicxx (found version "3.1")
-- Found MPI: TRUE (found version "3.1") found components:  CXX
-- Could NOT find Unwind (missing: Unwind_LIBRARIES)
-- Found Sphinx: /usr/bin/sphinx-build
-- Configuring done
-- Generating done
-- Build files have been written to: /local/scratch/tmp/ARBOR/nsuite/build/arbor/build
Scanning dependencies of target default_catalogue_cpp_target


[ 38%] Building CXX object arbor/CMakeFiles/arbor.dir/cell_group_factory.cpp.o
In file included from /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/simd/native.hpp:62:0,
                 from /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/simd/simd.hpp:9,
                 from /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/backends/multicore/shared_state.cpp:15:
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/simd/avx512.hpp: In static member function ‘static __mmask8 arb::simd::simd_detail::avx512_mask8::broadcast(bool)’:
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/simd/avx512.hpp:52:34: error: ‘_mm512_int2mask’ was not declared in this scope
         return _mm512_int2mask(-b);
                                  ^
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/simd/avx512.hpp: In static member function ‘static __mmask8 arb::simd::simd_detail::avx512_mask8::copy_from(const bool*)’:
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/simd/avx512.hpp:72:55: error: ‘_mm512_int2mask’ was not declared in this scope
         return _mm512_int2mask(_mm256_movemask_epi8(a));
                                                       ^
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/simd/avx512.hpp: In static member function ‘static bool arb::simd::simd_detail::avx512_mask8::element(const __mmask8&, int)’:
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/simd/avx512.hpp:172:33: error: ‘_mm512_mask2int’ was not declared in this scope
         return _mm512_mask2int(k)&(1<<i);
                                 ^
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/simd/avx512.hpp: In static member function ‘static void arb::simd::simd_detail::avx512_mask8::set_element(__mmask8&, int, bool)’:
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/simd/avx512.hpp:176:34: error: ‘_mm512_mask2int’ was not declared in this scope
         int n = _mm512_mask2int(k);
                                  ^
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/simd/avx512.hpp:177:47: error: ‘_mm512_int2mask’ was not declared in this scope
         k = _mm512_int2mask((n&~(1<<i))|(b<<i));
                                               ^
 /local/scratch/tmp/ARBOR/nsuite/build/arbor/arbor/include/arbor/simd/avx512.hpp: In static member function ‘static __mmask8 arb::simd::simd_detail::avx512_mask8::mask_unpack(long long unsigned int)’:
```

The CMake configure line is : 

```
==  ARBOR: repository has already downloaded
==  ARBOR: cmake -DCMAKE_INSTALL_PREFIX:PATH= /local/scratch/tmp/ARBOR/nsuite/install -DARB_WITH_MPI=ON -DARB_WITH_GPU=OFF -DARB_ARCH=native -DARB_VECTORIZE=ON
==  ARBOR: build
== ERROR see /local/scratch/tmp/ARBOR/nsuite/build/log_arbor
```
No specific module/env setup except loading `gcc intel-parallel-studio/cluster.2018.3 cmake python3` etc. I hope I am not missing anything obvious.

**Comment by bcumming** [comment]
Hello Pramod.

It looks like we have three separate problems here.

## problem 1: Intel compiler

Arbor doesn't support the Intel compiler. It isn't required for vectorization, because Arbor uses intrinsics.

https://arbor.readthedocs.io/en/latest/install.html#compilers

You can use different compilers to compile the different simulatiors in nsuite, e.g. if you want to compiler Arbor and NEURON with gcc and CoreNeuron with Intel:

```
./install-local.sh arbor neuron
./install-local.sh coreneuron -e my_special_intel_setup.sh
```

When you run `run-bench.sh` afterwards, it will make sure that the correct environment is configured for each simulator.

## problem 2: gcc 7

I don't know exactly what the problem is here. Can you please tell me which version of cuda toolkit and which MPI distribution you are using?

The build script for arbor in nsuite is building the examples. which it shouldn't have to do anymore. Could you change the following line

```bash
// change this line
make -j $ns_makej examples &>> "$out"
// to the following
make -j $ns_makej &>> "$out"
```

## problem 3: gcc 6

Which CPU architecture are you using? If it is SkyLake-512 then the issue might be that gcc6 doesn't support some of the intrinsics that Arbor uses when SkyLake-512 is detected with `-march=native`... but I don't really have enough information. Do you have more information about your system?

**Comment by bcumming** [comment]
Have you followed up on this Pramod? Let us know if you want more help.

**Comment by pramodk** [comment]
Sorry for delay, didn't get time to test this further.

> problem 1: Intel compiler
> Arbor doesn't support the Intel compiler. It isn't required for vectorization, because Arbor uses intrinsics.

Ah ok. I used Intel because it was just easy to configure all builds. I will make necessary changes.

> I don't know exactly what the problem is here. Can you please tell me which MPI distribution you are using?

This is with HPE MPI (SGI-MPT) on BB5 system.

> Which CPU architecture are you using? If it is SkyLake-512 then the issue might be that gcc6 doesn't support some of the intrinsics that Arbor uses when SkyLake-512 is detected with -march=native

Yes, its SKL-512 (Xeon 6140). I saw in the documentation GCC 6 is supported and hence GCC/6.3 (which was just default module on the system).

I will re-test `problem 2: gcc 7` and update this soon.

**Comment by pramodk** [comment]
I tested with GCC/7.3.0 and it builds fine. With GCC/6.3.0, I get following : 

```
$ module load cmake/3.12.0 hpe-mpi gcc/6.4.0 python-dev/0.1/python3
Autoloading python/3.6.5
Autoloading hdf5/1.10.4/serial
$ cmake .. -DARB_WITH_MPI=ON -DARB_WITH_GPU=OFF -DARB_ARCH=native -DARB_VECTORIZE=ON
-- The C compiler identification is GNU 6.4.0
-- The CXX compiler identification is GNU 6.4.0
-- Check for working C compiler: /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/gcc-6.4.0-i6lyqfscua/bin/gcc
-- Check for working C compiler: /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/gcc-6.4.0-i6lyqfscua/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/gcc-6.4.0-i6lyqfscua/bin/g++
-- Check for working CXX compiler: /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/gcc-6.4.0-i6lyqfscua/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found MPI_CXX: /lib64/libpthread.so (found version "3.1")
-- Found MPI: TRUE (found version "3.1") found components:  CXX
-- Could NOT find Unwind (missing: Unwind_LIBRARIES)
-- Found Sphinx: /usr/bin/sphinx-build
-- Configuring done
-- Generating done
-- Build files have been written to: /gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/build
$ make VERBOSE=1
…..
[ 34%] Building CXX object arbor/CMakeFiles/arbor.dir/backends/multicore/mechanism.cpp.o
cd /gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/build/arbor && /gpfs/bbp.cscs.ch/apps/hpc/jenkins/deploy/compilers/2018-12-18/linux-rhel7-x86_64/gcc-4.8.5/gcc-6.4.0-i6lyqfscua/bin/g++  -DARB_HAVE_MPI -DMPICH_SKIP_MPICXX=1 -DOMPI_SKIP_MPICXX=1 -I/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/arbor -I/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/arbor/include -I/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/build/arbor/include -isystem /opt/hpe/hpc/mpt/mpt-2.16/include  -O3 -DNDEBUG   -g -Wall -Wno-maybe-uninitialized -march=native -std=gnu++14 -o CMakeFiles/arbor.dir/backends/multicore/mechanism.cpp.o -c /gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/arbor/backends/multicore/mechanism.cpp
In file included from /gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/arbor/include/arbor/simd/native.hpp:62:0,
                 from /gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/arbor/include/arbor/simd/simd.hpp:9,
                 from /gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/arbor/backends/multicore/partition_by_constraint.hpp:5,
                 from /gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/arbor/backends/multicore/mechanism.hpp:15,
                 from /gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/arbor/backends/multicore/mechanism.cpp:19:
/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/arbor/include/arbor/simd/avx512.hpp: In static member function ‘static __mmask8 arb::simd::detail::avx512_mask8::broadcast(bool)’:
/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/arbor/include/arbor/simd/avx512.hpp:52:34: error: ‘_mm512_int2mask’ was not declared in this scope
         return _mm512_int2mask(-b);
                                  ^
/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/arbor/include/arbor/simd/avx512.hpp: In static member function ‘static __mmask8 arb::simd::detail::avx512_mask8::copy_from(const bool*)’:
/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/ARBOR/arbor/arbor/include/arbor/simd/avx512.hpp:72:55: error: ‘_mm512_int2mask’ was not declared in this scope
         return _mm512_int2mask(_mm256_movemask_epi8(a));
```
As mentioned in the previous comment, it's `Intel Xeon Gold 6140 CPU`.

If I should test something for `GCC/6.3.0`, let me know. If Intel compiler is not supported, may be at cmake stage we can throw an error rather than failing later during build stage? 

I have working build so its ok if we close this issue.

**Comment by bcumming** [comment]
Hi Pramod, I think that GCC 6 didn't support the `_mm512_int2mask` instrinsic, and support was added in GCC 7. See the link below.

https://www.mail-archive.com/gcc-bugs@gcc.gnu.org/msg529680.html

I would suggest trying to target Broadwell (i.e. AVX2) with `ARB_ARCH=broadwell`, even though you are using a processor that supports AVX512.

We have found that for many models, AVX2 is faster on SkyLake than AVX512 for two reasons:
* Arbor reorders loops inside `current`  and `state` mechanism updates so that it can optimise for contiguous and constant indexes for loading/storing variables like voltage and currents: if using vectors of width N, if there are N contiguous indexes Arbor explicitly uses a vector load, and if there are N repeated indexes, a single load is performed. This avoids explicit gather and scatter intrinsics, which as you know are really expensive. Wider vector lanes are often slower because there is a higher proportion of gather-scatter than vector loads and stores, e.g. there are less likely to be 8 contiguous or constant indexes in a row than 4.
* When using AVX512 the processor runs at a lower frequency.
